keps/sig-api-machinery/4006-transition-spdy-to-websockets/README.md

```diff
@@ -0,0 +1,1200 @@
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Some Kubernetes clients need to communicate with the API Server using a bi-directional
+streaming protocol, instead of the standard HTTP request/response mechanism. A streaming
+protocol provides the ability read and write arbitrary data messages between the
+client and server, instead of providing a single response to a client request.
+For example, the commands `kubectl exec`, `kubectl attach`, `kubectl port-forward`,
+and `kubectl cp` all benefit from a bi-directional streaming protocol. Currently,
+the bi-directional streaming solution for these `kubectl` commands is SPDY/3.1. For
+the communication leg between `kubectl` and the API Server, this enhancement transitions
+the bi-directional streaming protocol to WebSockets from SPDY/3.1 for three `kubectl`
+commands: `exec`, `attach`, and `cp`.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+The SPDY streaming protocol has been deprecated for around eight years, and by now
+many proxies, gateways, and load-balancers do not support SPDY. Our effort to modernize
+the streaming protocol between Kubernetes clients and the API Server is necessary
+to enable the aforementioned intermediaries. Additionally, the better supported and
+modern WebSockets should be more reliable and maintainable than SPDY.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+1. Transition the bi-directional streaming protocol from SPDY/3.1 to WebSockets for
+`kubectl exec`, `kubectl attach`, and `kubectl cp` for the communication leg
+between `kubectl` and the API Server.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+1. We do not intend to initially transition the current SPDY bi-directional streaming
+protocol for `kubectl port-forward`. This command requires a different subprotocol
+than the previously stated three `kubectl` commands (which use the `RemoteCommand`
+subprotocol). We intend to transition `kubectl port-forward` in a future release.
+
+2. We do not initially intend to modify *any* of the SPDY communication paths between
+other cluster components other than `kubectl` and the API Server. In other words,
+the current SPDY communication paths between the API Server and Kubelet, or Kubelet
+and the Container Runtime will not change.
```

Comment:
I'd like enough design detail to show that we will be able to switch `kubectl port-forward` to websockets in the future.  Incremental progress toward websockets is great. Having two protocols indefinitely is an outcome I want to avoid.